### PR TITLE
Gh-18 hierarchical policy

### PR DESCRIPTION
**Pull request by ac74475**

### CODE REVIEW

File: service-impl/policy-service-impl/hierarchical-policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/impl/HierarchicalPolicyService.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+import java.io.IOException;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Objects;
+import java.util.concurrent.CompletableFuture;
+import java.util.stream.Collectors;
+
+/**
+ * By having the policies stored in several key value stores we can attach policies
+ * at either the resource, data type.
```
Comment (m09526) [inline service-impl/policy-service-impl/hierarchical-policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/impl/HierarchicalPolicyService.java]: at either the resource or data type level.

File: service-impl/policy-service-impl/hierarchical-policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/impl/HierarchicalPolicyService.java [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+    private HashMap<String, Policy> dataTypePoliciesMap;
+    private HashMap<Resource, Policy> resourcePoliciesMap;
+
+    public HierarchicalPolicyService() {
+        this.dataTypePoliciesMap = new HashMap<>();
+        this.resourcePoliciesMap = new HashMap<>();
+    }
+
+    public HierarchicalPolicyService(final HashMap<String, Policy> dataTypePoliciesMap, final HashMap<Resource, Policy> resourcePoliciesMap) {
```
Comment (m09526) [inline service-impl/policy-service-impl/hierarchical-policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/impl/HierarchicalPolicyService.java]: Null checks in here please!

File: service/policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/Policy.java [hunk 3]
```diff
@@ -16,85 +16,155 @@
...
+
+    public Rules<Resource> getResourceRules() {
+        return resourceRules;
+    }
+
+    public void setResourceRules(final Rules<Resource> resourceRules) {
+        this.resourceRules = resourceRules;
+    }
+
+    private String generateUUID() {
```
Comment (m09526) [inline service/policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/Policy.java]: This should be static

File: service/policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/Policy.java [hunk 4]
```diff
@@ -16,85 +16,155 @@
...
+        return UUID.randomUUID().toString();
+    }
+
+    private void addMessage(final String newMessage, final Rules rules) {
+        String currentMessage = rules.getMessage();
+        if (currentMessage == null || currentMessage.equals("")) {
```
Comment (m09526) [inline service/policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/Policy.java]: Check string is empty with String.isEmpty()

File: service/policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/Policy.java [hunk 5]
```diff
@@ -16,85 +16,155 @@
...
+        resourceRules.simpleFunctionRule(generateUUID(), rule);
+        addMessage(message, resourceRules);
+        return this;
+    }
+
+    public User getOwner() {
+        return owner;
+    }
+
+    public void setOwner(final User owner) {
```
Comment (m09526) [inline service/policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/Policy.java]: null check?

File: service-impl/policy-service-impl/hierarchical-policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/impl/HierarchicalPolicyService.java (hunk 2 above)
Comment (ac74475) [reply service-impl/policy-service-impl/hierarchical-policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/impl/HierarchicalPolicyService.java]: added the null checks

File: service-impl/policy-service-impl/hierarchical-policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/impl/HierarchicalPolicyService.java [hunk 6]
```diff
@@ -0,0 +1,161 @@
...
+            } else {
+                inheritedRules = inheritedPolicy.getRecordRules();
+            }
+        }
+        Policy newPolicy = resourcePoliciesMap.getOrDefault(resource, null);
+        if (newPolicy == null) {
+            return inheritedRules;
+        } else {
+            Rules<T> newRules;
+            if (canAccessRequest) {
```
[the remark this answers is not in the document]
Comment (ac74475) [reply service-impl/policy-service-impl/hierarchical-policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/impl/HierarchicalPolicyService.java]: added javadoc to the method

File: service/policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/Policy.java (hunk 5 above)
Comment (ac74475) [reply service/policy-service/src/main/java/uk/gov/gchq/palisade/policy/service/Policy.java]: added the null check

**Comment by m09526** [comment]
Approved and merged into develop

**Comment by m09526** [comment]
Merged

**Pull request merged**